analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/LoopDetector.cs

```diff
@@ -0,0 +1,92 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.CFG.Roslyn;
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn;
+
+internal class LoopDetector
+{
+    private readonly HashSet<int> loopBlock;
+
+    public LoopDetector(ControlFlowGraph cfg)
+    {
+        loopBlock = DetectLoopBlockOrdinals(cfg).ToHashSet();
+    }
+
+    public bool IsInLoop(BasicBlock block) =>
+        loopBlock.Contains(block.Ordinal);
+
+    // Detects loops in the cfg using a modified DFS algorithm:
+    // Explore the cfg depth-first, keeping track of the path from the root to the current block.
+    // If the current block is already in the path, a loop is detected.
+    // Otherwise, if the last block in the path is part of an already detected loop that intersects with an earlier part of the path, merge the relevant sub-path with the loop.
+    // Due to the nature of DFS:
+    // If a block is processed already, current path has no loop and is not intersecting with already detected loops, the block is not part of any loop.
+    private static IEnumerable<int> DetectLoopBlockOrdinals(ControlFlowGraph cfg)
+    {
+        var processed = new HashSet<int>();
+        var loops = new List<HashSet<int>>();
+        var toProcess = new Stack<List<int>>();
+        toProcess.Push([1]);
+        while (toProcess.TryPop(out var path))
+        {
+            var last = path[path.Count - 1];
+            if (processed.Add(last))
+            {
+                ScheduleSuccessors(path, last);
+            }
+            else
+            {
+                var index = path.IndexOf(last);
+                if (index < path.Count - 1)         // is the block in the path twice? -> loop
+                {
+                    loops.Add(path.GetRange(index, path.Count - 1 - index).ToHashSet());    // equivalent to [index..^1]
+                }
+                else
+                {
+                    MergeWithIntersectingLoops(path, last);
+                }
+            }
+        }
+        return loops.SelectMany(x => x);
+
+        void ScheduleSuccessors(List<int> path, int last)
+        {
+            foreach (var successor in cfg.Blocks[last].SuccessorBlocks.Select(x => x.Ordinal))
+            {
+                toProcess.Push([.. path, successor]);   // add successor to the path and schedule it for processing
+            }
+        }
+
+        // For a given path [..A..B], if we find a loop that both A and B are part of, all blocks between A and B should also be considered part of that loop.
+        void MergeWithIntersectingLoops(List<int> path, int last)
+        {
+            foreach (var loop in loops.Where(x => x.Contains(last)))    // B is part of the loop
```

Comment:
I tried to make the `loops` a simple hashset and no UT broke.
So we either miss an UT (I'm consider some parallel goto between two loops like `---O===O---` scenario, or it is really pointless to have list of hashsets

Reply:
Yes, we were missing a UT. I added it.